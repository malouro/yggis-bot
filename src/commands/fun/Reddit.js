import ParseFeed from 'feed-read-parser';

import Command from '../../classes/Command';

export default class Reddit extends Command {
	constructor({ t } = { t: () => null }) {
		super({
			name: 'Reddit',
			category: 'fun',
			description: 'Grab a random Reddit post from a specific subreddit.',
			usage: {
				args: [
					{
						name: 'subreddit',
						description: 'The subreddit to grab a post from.',
						chainable: false,
						type: 'string',
					},
				],
			},
		});

		this.t = t;
		this.translationKey = 'reddit';

		this.name = this.t('COMMANDS', this.translationKey, 'name') || this.name;
		this.description =
			this.t('COMMANDS', this.translationKey, 'description') ||
			this.description;
	}

	preAction({ args }) {
		this.subreddit = args[1] ? args[1].toLocaleLowerCase() : 'all';
		return null;
	}

	action({ message, logger }) {
		const { subreddit } = this;
		const redditUrl = `https://www.reddit.com/r/${subreddit}/.rss`;
		let messageContent = '';

		ParseFeed(redditUrl, (error, articles) => {
			if (error) {
				return logger.debug.log({
					level: 'error',
					message: error,
				});
			}

			if (articles.length === 0) {
				const invalidMessage =
					(typeof this.t(
						'COMMANDS',
						this.translationKey,
						'invalidSubreddit'
					) === 'function' &&
						this.t(
							'COMMANDS',
							this.translationKey,
							'invalidSubreddit'
						)(subreddit)) ||
					`Oops! The subreddit \`/r/${subreddit}\` is either empty or non-existent.`;

				return message.reply(invalidMessage);
			}

			const links = [];
			const titles = [];
			const authors = [];
			const getImage = content => {
				const regex = /<a .*?href="(.*?)">/gi;
				const matches = [...content.matchAll(regex)];

				return matches[2][1].includes('i.redd.it') ?
					matches[2][1] :
					matches[3][1];
			};

			articles.forEach(article => {
				titles.push(article.title);
				authors.push(article.author);
				links.push(article.link);
			});

			const rng = Math.floor(Math.random() * links.length);

			messageContent +=
				(typeof this.t('COMMANDS', this.translationKey, 'postedByText') ===
					'function' &&
					this.t(
						'COMMANDS',
						this.translationKey,
						'postedByText'
					)(titles[rng], authors[rng])) ||
				`**${titles[rng]}** posted by ${authors[rng]}`;

			if (articles[rng].content.includes('img src')) {
				const image = getImage(articles[rng].content);

				messageContent += `\n${image}`;
			}

			messageContent += `\n<${links[rng]}>`;

			return message.channel.send(messageContent);
		});
	}
}
