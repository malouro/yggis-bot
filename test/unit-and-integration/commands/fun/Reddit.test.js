import Reddit from '../../../../src/commands/fun/Reddit';

import { runCommand, MockTranslateFunc } from '../../../testHelpers';

jest.mock('feed-read-parser', () => (url, callback) => {
	switch (url) {
	/* Forcing a return of an error */
	case 'https://www.reddit.com/r/FORCE_ERROR/.rss':
		return callback(new Error());

	/* /r/all */
	case 'https://www.reddit.com/r/all/.rss':
		return callback(null, [{
			title: 'POST_TITLE',
			// Imitate XML that can return from feed-read-parser
			content: `<table> <tr><td> <a href="https://www.reddit.com/r/pics/comments/SOME_POST/">
			<img src="https://b.thumbs.redditmedia.com/yjc-IUtDVd5xHZkeMjJ043eDEDpBMx0CGSbrQ9CsbGk.jpg"
			alt="ðŸ’ª" title="ðŸ’ª" /> </a> </td><td> &#32; submitted by &#32;
			<a href="https://www.reddit.com/user/SOME_AUTHOR"> /u/SOME_AUTHOR </a>
			&#32; to &#32; <a href="https://www.reddit.com/r/all/"> r/all </a> <br/>
			<span><a href="https://i.redd.it/wqd0hx650k651.jpg">[link]</a></span> &#32; <span>
			<a href="https://www.reddit.com/r/all/comments/SOME_POST/">[comments]</a></span> </td></tr></table>`,
			published: 'SOME_DATE',
			author: '/u/SOME_AUTHOR',
			link: 'https://www.reddit.com/r/all/comments/SOME_POST/',
			feed: {
				source: 'https://www.reddit.com/r/all/.rss',
				link: 'https://www.reddit.com/r/all/',
				name: 'all subreddits'
			}
		}]);

	// ???
	default:
		break;
	}
});

describe('Reddit command', () => {
	const RedditCommand = new Reddit({ t: MockTranslateFunc() });

	test('runs with no args and grabs a post from r/all', async () => {
		const result = await runCommand(RedditCommand, { args: [] });

		expect(result.message.channel.send.mock.results[0].value).toMatchSnapshot();
	});

	test('logs any errors from parsing Reddit feed', async () => {
		const result = await runCommand(RedditCommand, { args: [ '!reddit', 'FORCE_ERROR' ] });

		expect(result.message.channel.send.mock.results[0].value).toMatchSnapshot();
	});
});
