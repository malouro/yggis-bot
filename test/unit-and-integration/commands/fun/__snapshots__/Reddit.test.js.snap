// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reddit command runs with no args and grabs a post from r/all 1`] = `
"**POST_TITLE**
posted by /u/SOME_AUTHOR
https://www.reddit.com/r/all/
<https://www.reddit.com/r/all/comments/SOME_POST/>"
`;
